dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/SyncNamesWithDqtJob.cs

```diff
@@ -0,0 +1,102 @@
+using Microsoft.EntityFrameworkCore;
+using Polly;
+using TeacherIdentity.AuthServer.Events;
+using TeacherIdentity.AuthServer.Infrastructure.Http;
+using TeacherIdentity.AuthServer.Models;
+using TeacherIdentity.AuthServer.Services.DqtApi;
+
+namespace TeacherIdentity.AuthServer.Jobs;
+
+public class SyncNamesWithDqtJob : IDisposable
+{
+    private const int DqtApiRetryCount = 3;
+
+    private readonly TimeSpan DqtApiDefaultRetryAfter = TimeSpan.FromSeconds(30);
+    private readonly TeacherIdentityServerDbContext _readDbContext;
+    private readonly TeacherIdentityServerDbContext _writeDbContext;
+    private readonly IDqtApiClient _dqtApiClient;
+    private readonly IClock _clock;
+
+    public SyncNamesWithDqtJob(
+        IDbContextFactory<TeacherIdentityServerDbContext> dbContextFactory,
+        IDqtApiClient dqtApiClient,
+        IClock clock)
+    {
+        _readDbContext = dbContextFactory.CreateDbContext();
+        _writeDbContext = dbContextFactory.CreateDbContext();
+        _dqtApiClient = dqtApiClient;
+        _clock = clock;
+    }
+
+    public async Task Execute(CancellationToken cancellationToken)
+    {
+        await foreach (var userWithTrn in _readDbContext.Users.Where(u => u.Trn != null).AsNoTracking().AsAsyncEnumerable())
+        {
+            var dqtTeacher = await Policy
+                    .Handle<TooManyRequestsException>()
+                    .WaitAndRetryAsync(
+                        DqtApiRetryCount,
+                        sleepDurationProvider: (retryCount, exception, context) =>
+                        {
+                            var rateLimitingException = exception as TooManyRequestsException;
+                            return rateLimitingException!.RetryAfter ?? DqtApiDefaultRetryAfter;
+                        },
+                        onRetryAsync: (exception, delay, retryCount, context) =>
+                        {
+                            return Task.CompletedTask;
+                        })
+                    .ExecuteAsync(() => _dqtApiClient.GetTeacherByTrn(userWithTrn.Trn!, cancellationToken));
+
+            if (dqtTeacher is null)
+            {
+                continue;
+            }
+
+            var changes = UserUpdatedEventChanges.None |
+                        (userWithTrn.FirstName != dqtTeacher!.FirstName ? UserUpdatedEventChanges.FirstName : UserUpdatedEventChanges.None) |
+                        (userWithTrn.MiddleName != dqtTeacher!.MiddleName ? UserUpdatedEventChanges.MiddleName : UserUpdatedEventChanges.None) |
```

Comment:
We may have some `null` middle names (while the DQT API will always return empty strings)
```suggestion
                        ((userWithTrn.MiddleName ?? string.Empty) != dqtTeacher!.MiddleName ? UserUpdatedEventChanges.MiddleName : UserUpdatedEventChanges.None) |
```

Reply:
Good point .. I did mean to check about that